ext/cgi/escape/escape.c

```diff
@@ -11,26 +11,26 @@ RUBY_EXTERN const signed char ruby_digit36_to_number_table[];
 static VALUE rb_cCGI, rb_mUtil, rb_mEscape;
 static ID id_accept_charset;
 
+static const char* html_escape_table[UCHAR_MAX+1] = { NULL };
+static size_t html_escape_len[UCHAR_MAX+1] = { 0 };
+static size_t html_escape_max_len = 0;
+
 static void
-html_escaped_cat(VALUE str, char c)
+init_html_escape_tables(void)
 {
-    switch (c) {
-      case '\'':
-	rb_str_cat_cstr(str, "&#39;");
-	break;
-      case '&':
-	rb_str_cat_cstr(str, "&amp;");
-	break;
-      case '"':
-	rb_str_cat_cstr(str, "&quot;");
-	break;
-      case '<':
-	rb_str_cat_cstr(str, "&lt;");
-	break;
-      case '>':
-	rb_str_cat_cstr(str, "&gt;");
-	break;
-    }
+#define HTML_ESCAPE(c, str) do { \
+    html_escape_table[c] = str; \
```

Comment:
Possibly, define variable `int len = strlen(str);` here, and use it following?

Reply:
That might be robust and easier to read. I did so https://github.com/ruby/ruby/pull/2226/commits/9bb706a899880c3563c1c3e329ad061e9a4716ee